pkg/volume/volume_linux.go

```diff
@@ -36,60 +39,187 @@ const (
 // SetVolumeOwnership modifies the given volume to be owned by
 // fsGroup, and sets SetGid so that newly created files are owned by
 // fsGroup. If fsGroup is nil nothing is done.
-func SetVolumeOwnership(mounter Mounter, fsGroup *int64) error {
-
+func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy) error {
 	if fsGroup == nil {
 		return nil
 	}
 
+	fsGroupPolicyEnabled := utilfeature.DefaultFeatureGate.Enabled(features.ConfigurableFSGroupPolicy)
+
 	klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
 
-	return filepath.Walk(mounter.GetPath(), func(path string, info os.FileInfo, err error) error {
+	// This code exists for legacy purposes, so as old behaviour is entirely preserved when feature gate is disabled
+	// TODO: remove this when ConfigurableFSGroupPolicy turns beta.
+	if !fsGroupPolicyEnabled {
+		return legacyPermissionChange(mounter, fsGroup)
+	}
+
+	if skipPermissionChange(mounter, fsGroup, fsGroupChangePolicy) {
+		klog.V(3).Infof("skipping permission and ownership change for volume %s", mounter.GetPath())
+		return nil
+	}
+
+	return walkDeep(mounter.GetPath(), func(path string, info os.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
+		return changeFilePermission(path, fsGroup, mounter.GetAttributes().ReadOnly, info)
+	})
 
-		// chown and chmod pass through to the underlying file for symlinks.
-		// Symlinks have a mode of 777 but this really doesn't mean anything.
-		// The permissions of the underlying file are what matter.
-		// However, if one reads the mode of a symlink then chmods the symlink
-		// with that mode, it changes the mode of the underlying file, overridden
-		// the defaultMode and permissions initialized by the volume plugin, which
-		// is not what we want; thus, we skip chown/chmod for symlinks.
-		if info.Mode()&os.ModeSymlink != 0 {
-			return nil
-		}
+}
 
-		stat, ok := info.Sys().(*syscall.Stat_t)
-		if !ok {
-			return nil
+func legacyPermissionChange(mounter Mounter, fsGroup *int64) error {
```

Comment:
I feel SettingOwnership is better than PermissionChange? This is mainly for change the ownership, not the file permission.

Reply:
it changes both. Not just ownership.  but we are just talking about function name right?

Reply:
we call it SetVolumeOwnership because the main goal is change the ownership. I think we don't intentionally change permission here, just want to make sure user/group has read/write permission so that container can access files when it run as user.

Reply:
why do you say we do not "intentionally change permission"? because we do  -https://github.com/kubernetes/kubernetes/blob/223ef76db5a4a537ab670a2f66a830ee97c89105/pkg/volume/volume_linux.go#L88

Reply:
@jingxu97 given a volume with messed up permissions, if we only performed chown and gave ownership to the `fsgroup`, that alone will not be sufficient for user to read/write files from the volume unless a `chmod` is also performed.

Reply:
I am thinking we have another field "DefaultMode" for some volumes to change permission bits. That's different from changing ownership here. But I agree changing ownership and permission is very related. 

About the naming, just want to reduce confusion. But I am ok if you prefer this one.